File: 0589-N-ary Tree Preorder Traversal/0589-N-ary Tree Preorder Traversal.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> result;
        if (root == nullptr) {
            return result;
        }
        stack<Node*> St;
        St.push(root);
        while (!St.empty()) {
            Node* node = St.top();
            St.pop();
            result.push_back(node->val);
            int n = node->children.size();
            for (int i = n - 1; i >= 0; --i) {
                St.push(node->children[i]);
            }
        }
        return result;
    }
};

File: 0501-0600/0589-N-ary Tree Preorder Traversal/0589-N-ary Tree Preorder Traversal.py
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if root is None:
            return []
        St = [root]
        result = []
        while St:
            node = St.pop()
            result.append(node.val)
            St += node.children[::-1]
        return result
